File: rerun_cpp/src/rerun/archetypes/asset_video_ext.cpp
#include <algorithm>
#include <fstream>
#include <string>

#include "../c/rerun.h"
#include "../string_utils.hpp"
#include "asset_video.hpp"

#include <arrow/array/array_binary.h>
#include <arrow/array/array_nested.h>

// It's undefined behavior to pre-declare std types, see http://www.gotw.ca/gotw/034.htm
// We want to use `std::filesystem::path`, so we have it include it in the header.
// <CODEGEN_COPY_TO_HEADER>

#include <chrono>
#include <filesystem>

// </CODEGEN_COPY_TO_HEADER>

namespace rerun::archetypes {

#if 0
        // <CODEGEN_COPY_TO_HEADER>

        /// Creates a new `AssetVideo` from the file contents at `path`.
        ///
        /// The `MediaType` will be guessed from the file extension.
        ///
        /// If no `MediaType` can be guessed at the moment, the Rerun Viewer will try to guess one
        /// from the data at render-time. If it can't, rendering will fail with an error.
        static Result<AssetVideo> from_file(const std::filesystem::path& path);

        /// Creates a new `AssetVideo` from the given `bytes`.
        ///
        /// If no `MediaType` is specified, the Rerun Viewer will try to guess one from the data
        /// at render-time. If it can't, rendering will fail with an error.
        static AssetVideo from_bytes(
            rerun::Collection<uint8_t> bytes, std::optional<rerun::components::MediaType> media_type = {}
        ) {
            AssetVideo asset = AssetVideo(std::move(bytes));
            // TODO(jan): we could try and guess using magic bytes here, like rust does.
            if (media_type.has_value()) {
                return std::move(asset).with_media_type(media_type.value());
            }
            return asset;
        }

        /// Determines the presentation timestamps of all frames inside the video.
        ///
        /// Returned timestamps are in nanoseconds since start and are guaranteed to be monotonically increasing.
        Result<std::vector<std::chrono::nanoseconds>> read_frame_timestamps_nanos() const;

        /// DEPRECATED: Use `read_frame_timestamps_nanos` instead.
        [[deprecated("Renamed to `read_frame_timestamps_nanos`"
        )]] Result<std::vector<std::chrono::nanoseconds>>
            read_frame_timestamps_ns() const {
            return read_frame_timestamps_nanos();
        }

        // </CODEGEN_COPY_TO_HEADER>
#endif

    Result<AssetVideo> AssetVideo::from_file(const std::filesystem::path& path) {
        std::ifstream file(path, std::ios::binary);
        if (!file) {
            return Error(ErrorCode::FileOpenFailure, "Failed to open file: " + path.string());
        }

        file.seekg(0, std::ios::end);
        std::streampos length = file.tellg();
        file.seekg(0, std::ios::beg);

        std::vector<uint8_t> data(static_cast<size_t>(length));
        file.read(reinterpret_cast<char*>(data.data()), length);

        return AssetVideo::from_bytes(
            Collection<uint8_t>::take_ownership(std::move(data)),
            rerun::components::MediaType::guess_from_path(path)
        );
    }

    static int64_t* alloc_timestamps(void* alloc_context, uint32_t num_timestamps) {
        auto frame_timestamps_ptr =
            static_cast<std::vector<std::chrono::nanoseconds>*>(alloc_context);
        frame_timestamps_ptr->resize(num_timestamps);
        return reinterpret_cast<int64_t*>(frame_timestamps_ptr->data());
    }

    Result<std::vector<std::chrono::nanoseconds>> AssetVideo::read_frame_timestamps_nanos() const {
        static_assert(sizeof(int64_t) == sizeof(std::chrono::nanoseconds::rep));

        if (!blob.has_value()) {
            return std::vector<std::chrono::nanoseconds>();
        }
        auto blob_list_array = std::dynamic_pointer_cast<arrow::ListArray>(blob.value().array);
        if (!blob_list_array) {
            return Error(ErrorCode::InvalidComponent, "Blob array is not a primitive array");
        }
        auto blob_array =
            std::dynamic_pointer_cast<arrow::PrimitiveArray>(blob_list_array->values());
        if (!blob_array) {
            return Error(ErrorCode::InvalidComponent, "Blob array is not a primitive array");
        }
        auto blob_array_data = blob_array->values();

        rr_string media_type_c = detail::to_rr_string(std::nullopt);
        if (media_type.has_value()) {
            auto media_type_array =
                std::dynamic_pointer_cast<arrow::StringArray>(media_type.value().array);
            if (media_type_array) {
                media_type_c = detail::to_rr_string(media_type_array->GetView(0));
            }
        }

        std::vector<std::chrono::nanoseconds> frame_timestamps;

        rr_error status = {};
        rr_video_asset_read_frame_timestamps_nanos(
            blob_array_data->data(),
            static_cast<uint64_t>(blob_array_data->size()),
            media_type_c,
            &frame_timestamps,
            &alloc_timestamps,
            &status
        );
        if (status.code != RR_ERROR_CODE_OK) {
            return Error(status);
        }

        return frame_timestamps;
    }

} // namespace rerun::archetypes

File: crates/store/re_types/src/archetypes/asset_video_ext.rs
use crate::components::MediaType;

use super::AssetVideo;

impl AssetVideo {
    /// Creates a new [`AssetVideo`] from the file contents at `path`.
    ///
    /// The [`MediaType`] will first be guessed from the file extension, then from the file
    /// contents if needed.
    ///
    /// If no [`MediaType`] can be guessed at the moment, the Rerun Viewer will try to guess one
    /// from the data at render-time. If it can't, rendering will fail with an error.
    ///
    /// Returns an error if the file cannot be read.
    #[cfg(not(target_arch = "wasm32"))]
    #[inline]
    pub fn from_file_path(filepath: impl AsRef<std::path::Path>) -> std::io::Result<Self> {
        let filepath = filepath.as_ref();
        let contents = std::fs::read(filepath)?;
        Ok(Self::from_file_contents(
            contents,
            MediaType::guess_from_path(filepath),
        ))
    }

    /// Creates a new [`AssetVideo`] from the given `contents`.
    ///
    /// The [`MediaType`] will be guessed from magic bytes in the data.
    ///
    /// If no [`MediaType`] can be guessed at the moment, the Rerun Viewer will try to guess one
    /// from the data at render-time. If it can't, rendering will fail with an error.
    #[inline]
    pub fn from_file_contents(contents: Vec<u8>, media_type: Option<impl Into<MediaType>>) -> Self {
        let media_type = media_type.map(Into::into);
        if let Some(media_type) = MediaType::or_guess_from_data(media_type, &contents) {
            Self::new(contents).with_media_type(media_type)
        } else {
            Self::new(contents)
        }
    }

    /// Determines the presentation timestamps of all frames inside the video.
    ///
    /// Returned timestamps are in nanoseconds since start and are guaranteed to be monotonically increasing.
    ///
    /// Panics if the serialized blob data doesn't have the right datatype.
    #[cfg(feature = "video")]
    pub fn read_frame_timestamps_nanos(&self) -> Result<Vec<i64>, re_video::VideoLoadError> {
        use re_types_core::Loggable as _;

        re_tracing::profile_function!();

        let Some(blob) = self.blob.as_ref() else {
            return Ok(Vec::new());
        };

        // Grab blob data without a copy.
        let blob_list_array = blob
            .array
            .as_any()
            .downcast_ref::<arrow::array::ListArray>()
            .expect("Video blob data is not a ListArray");
        let blob_data = blob_list_array.values().to_data();
        let blob_bytes = blob_data.buffer(0);

        let Some(media_type) = self
            .media_type
            .as_ref()
            .and_then(|mt| {
                MediaType::from_arrow(&mt.array)
                    .ok()
                    .and_then(|mt| mt.first().cloned())
            })
            .or_else(|| MediaType::guess_from_data(blob_bytes))
        else {
            return Err(re_video::VideoLoadError::UnrecognizedMimeType);
        };

        Ok(
            re_video::VideoData::load_from_bytes(blob_bytes, media_type.as_str())?
                .frame_timestamps_nanos()
                .collect(),
        )
    }

    /// DEPRECATED: renamed to `read_frame_timestamps_nanos`
    #[deprecated(since = "0.23.0", note = "Renamed to `read_frame_timestamps_nanos`")]
    #[cfg(feature = "video")]
    pub fn read_frame_timestamps_ns(&self) -> Result<Vec<i64>, re_video::VideoLoadError> {
        self.read_frame_timestamps_nanos()
    }
}
